workload/schemachanger: allow additional errors for ALTER COLUMN TYPE

spilchen opened the pull request:
This update enhances error handling for ALTER COLUMN TYPE operations, covering additional error cases that may arise:
- pgcode.CannotCoerce: Raised when a type conversion is not permissible (e.g., attempting an unsupported type change). We previously only did by inspecting the type change, but there are additional cases not covered in `schemachange.ClassifyConversion`. So, making it more general.
- pgcode.NumericValueOutOfRange: Raised when the new type cannot accommodate existing data values (e.g., converting from INT8 to INT2 where some values exceed the INT2 range).

I also moved existing potential errors up to the branch where we know we are moving to a known type. We don’t need to check for those if the type name doesn’t exist.

Epic: CRDB-25314
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/134962)
<!-- Reviewable:end -->

Dedej-Bergin reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134962)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @spilchen, and @vidit-bhat)

___
*[`pkg/workload/schemachange/operation_generator.go` line 2543 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134962#-OBVwnzP5MR-IXGnEUqT:-OBVwnzQ-YeochUCexrB:b-vx3ztu) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e26fb2209bb3109778adcab4d9361a1f5b4bfea/pkg/workload/schemachange/operation_generator.go#L2543)):*
> ```Go
> 
> 	stmt := makeOpStmt(OpStmtDDL)
> 	if newType != nil {
> ```

In your commit comment, you mentioned

"I also moved existing potential errors up to the branch where we know we
  are moving to a known type. We don’t need to check for those if the type
  name doesn’t exist."

What about if we have a pre-existing type name?  Wouldn't newType be nil but newTypeName would still have a value?

I was looking at the og.randType method and this seems to be possible.
![Screenshot 2024-11-12 at 10.35.51 AM.png](https://files.reviewableusercontent.io/images/reviewable/164547647/IbcAmIz4v03aXVYhhD9g9kXg/Screenshot%202024-11-12%20at%2010.35.51%E2%80%AFAM.png)

og.randTypeName inside of randType seems to find the pre-existing names with this query:
```
	var q = fmt.Sprintf(`
		SELECT schema, name
		    FROM [SHOW TYPES]
		   WHERE name LIKE '%s'
		ORDER BY random()
		   LIMIT 1;
		`, prefix+"%")
```


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
Thanks for taking a look Bergin.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134962#-:-OBWB0yY5LRqoGNRNyIk:b-7mcon2)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @Dedej-Bergin, and @vidit-bhat)

___
*[`pkg/workload/schemachange/operation_generator.go` line 2543 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134962#-OBVwnzP5MR-IXGnEUqT:-OBW7acy6bNGu2eSwfpw:b-wle1y3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e26fb2209bb3109778adcab4d9361a1f5b4bfea/pkg/workload/schemachange/operation_generator.go#L2543)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

In your commit comment, you mentioned

"I also moved existing potential errors up to the branch where we know we
  are moving to a known type. We don’t need to check for those if the type
  name doesn’t exist."

What about if we have a pre-existing type name?  Wouldn't newType be nil but newTypeName would still have a value?

I was looking at the og.randType method and this seems to be possible.
![Screenshot 2024-11-12 at 10.35.51 AM.png](https://files.reviewableusercontent.io/images/reviewable/164547647/IbcAmIz4v03aXVYhhD9g9kXg/Screenshot%202024-11-12%20at%2010.35.51%E2%80%AFAM.png)

og.randTypeName inside of randType seems to find the pre-existing names with this query:
```
	var q = fmt.Sprintf(`
		SELECT schema, name
		    FROM [SHOW TYPES]
		   WHERE name LIKE '%s'
		ORDER BY random()
		   LIMIT 1;
		`, prefix+"%")
```
</blockquote></details>

If there's an existing type name, then `newType` shouldn't be nil. It only becomes nil if, without other errors, we enter the `randTypeName` codepath, where a fake name is generated.
![image.png](https://files.reviewableusercontent.io/images/reviewable/171762/4XYuSy0_jy7Ts6Ynx9fAtepr/image.png)

The code snippet above shows this in line 3726. In that case, the code you referenced earlier would execute. However, if we reach the point where we query with `SHOW TYPES`, we should always get a valid result for `newType`.


<!-- Sent from Reviewable.io -->

spilchen committed:
workload/schemachanger: allow additional errors for ALTER COLUMN TYPE

This update enhances error handling for ALTER COLUMN TYPE operations,
covering additional error cases that may arise:
- pgcode.CannotCoerce: Raised when a type conversion is not permissible
(e.g., attempting an unsupported type change).
- pgcode.NumericValueOutOfRange: Raised when the new type cannot
accommodate existing data values (e.g., converting from INT8 to INT2
where some values exceed the INT2 range).

I also moved existing potential errors up to the branch where we know we
are moving to a known type. We don’t need to check for those if the type
name doesn’t exist.

Epic: CRDB-25314
Release note: none

Dedej-Bergin reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134962#-:-OBaFDhqCxLbYT9DcbNG:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DarrylWong and @vidit-bhat)



<!-- Sent from Reviewable.io -->
